Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -1342,12 +1340,12 @@ private void updateInformationFromServer(DocumentEvent e) {
         final int eventOffset = e.getOffset();
         final int deltaLength = e.getNewLength() - e.getOldLength();
         for (PluginNavigationRegion region : regions) {
-          if (region.offset <= eventOffset && eventOffset <= region.offset + region.length) {
-            region.length += deltaLength;
-          }
-          else if (region.offset >= eventOffset) {
+          if (eventOffset <= region.offset) {
             region.offset += deltaLength;
           }
+          else if (region.offset < eventOffset && eventOffset <= region.offset + region.length) {
+            region.length += deltaLength;
+          }
           for (PluginNavigationTarget target : region.getTargets()) {
             if (target.file.equals(filePath) && target.offset >= eventOffset) {
               target.offset += deltaLength;
```

Comment:
@scheglov I think if target region intersects with DocumentEvent range we should remove such target (and corresponding region if no more valid targets, or keep in mind that targets array may be empty). Agree?